File: it.unibo.bls19Local/src/main/java/it/unibo/bls/devices/DeviceFactory.java
package it.unibo.bls.devices;

import it.unibo.bls.devices.arduino.ButtonProxyArduino;
import it.unibo.bls.devices.arduino.LedProxyArduino;
import it.unibo.bls.devices.gui.ButtonAsGui;
import it.unibo.bls.devices.gui.LedAsGui;
import it.unibo.bls.devices.mock.ButtonMock;
import it.unibo.bls.devices.mock.LedMock;
import it.unibo.bls.interfaces.IButtonObservable;
import it.unibo.bls.interfaces.ILed;
import jssc.SerialPort;

public class DeviceFactory {

    public ILed createLed( LedType  devType )   {
        switch( devType ) {
            case LedMockObj:   return new LedMock();
            case LedGuiObj:    return LedAsGui.createLed();
            case LedOnArduino: return new LedProxyArduino(
                    DeviceConfig.serialPortNum, DeviceConfig.serialBaudrate );
            default: return null;
        }
    }
    public IButtonObservable createButton(ButtonType  devType )   {
        switch( devType ) {
            case ButtonMockObj: return new ButtonMock();
            case ButtonGuiObj: return  ButtonAsGui.createButton("CLICKME");
            case ButtonOnArduino: return new ButtonProxyArduino(
                    DeviceConfig.serialPortNum, DeviceConfig.serialBaudrate );
            default: return null;
        }
    }

}

File: it.unibo.bls19Local/src/main/kotlin/it/unibo/bls/kotlin/devices/DeviceFactory.kt
package it.unibo.bls.kotlin.devices

import it.unibo.bls.kotlin.devices.ButtonType
import it.unibo.bls.kotlin.devices.LedType
import it.unibo.bls.kotlin.devices.mock.LedMock
import it.unibo.bls.kotlin.interfaces.IButtonObservable
import it.unibo.bls.kotlin.interfaces.ILed
import it.unibo.bls.kotlin.devices.mock.ButtonMockCoroutine

class DeviceFactory {

    fun createLed(devType: LedType): ILed? {
        when (devType) {
            LedType.LedMockObj -> return LedMock()
            else -> return null
        }
    }

    fun createButton(devType: ButtonType): IButtonObservable? {
        when (devType) {
            ButtonType.ButtonMockObj -> return ButtonMockCoroutine()
            else -> return null
        }
    }

}
